### ISSUE TITLE
Subgrid turbulence scheme regression

### ISSUE THREAD

**Issue by fjansson** [dalesteam/dales#28]
Commit 2e6f725ff6f "Fix of Arnold: Fixed division-by-zero error in SGS-TKE fix" causes some runs to crash. In particular, the arm_brown test case crashes immediately.

That commit makes a change to avoid a division by 0, in case the horizontal velocity is exactly 0 (i.e. both u = 0 and v = 0). It also removes a cap on `local_dudz` (and same for `local_dvdz`) further down:
```
    if (abs(local_dudz) > abs(dudz(i,j))) then
        local_dudz = dudz(i,j)
    endif
```
It turns out this cap is needed when the subgrid-scale turbulence is low, i.e. `ekm(i,j,1)` is small.
This can happen either if the run is initialized with ekm = 0 at the surface as in the arm_brown case, or later in the run if low ekm from higher up in the atmosphere is advected down. A too large local_dudz will in turn cause instabilities later, by resulting in very large ekm.

I propose that this commit is reverted. Note that the horizontal velocity being exactly 0 is unlikely (for both components at once), and that tiny but non-zero velocities do not result in any problems. 

Also, the cap seems dubious. In practice it is triggered also in what seems to be normal circumstances. `local_dudz` and `dudz` may even have different signs, and then the sign of `local_dudz` is changed for what seems to be no physical reason.

**Comment by afmoene** [comment]
Dear all,

This is interesting. The fix was motivated by a number of purely 
convective runs that crashed. For that case the inclusion of a minimum 
value in the wind speed worked out well, and it removed at that time the 
need for the cap on local_dudz.  But apparently we need both the minimum 
on horv and some cap on local_dudz.

I'm a bit surprised that you get zero-values for ekm, because, if I 
recall well, the ekm values are set to a minimum value at initialization 
and so could/should not be zero. But if they are indeed zero or very 
small, it would be wise to put back the cap on local_dudz (but keep the 
minimum value for horv of 0.01 m/s).  On the unlikeliness of a zero wind 
speed: the code does allow initialization with zero wind speed, and that 
is what leads to crashes in convective simulations.

I agree that the physical basis of the value used for the cap is 
dubious. But this MOST-estimate of dudz was the only thing that I had 
around in the code as an estimate of the real dudz. Assuming that the 
cap was only needed to get through some initial unphysical phases of the 
simulation, I did not feel a strong objection to it. But any suggestion 
for a better estimate for the cap value is very welcome.

Cheers, Arnold

**Comment by fjansson** [comment]
There is a lower bound for ekm, e12min = 5.e-5, applied throughout the simulation. So we don't get zero for ekm - but small ekm cause large local_dudz, which results in a large tendency for ekm and later in a crash.

Putting the cap back but keeping the horv limit sounds good to me. If the model is initialized with 0 wind and with no initial random noise, there would indeed be a division by 0 otherwise.

**Comment by Chiil** [comment]
Hi Fredrik,

I have also dealt with this problem long time ago. The large tendency for ekm should solve itself as an increase in the diffusivity should smoothen out the field. It might be that your time step is at the edge of the diffusion criterion. What is the maximum (ekm*dt)/dx^2 that you permit?

Chiel

**Comment by fjansson** [comment]
Hi Chiel,

I left the peclet number on the default, 0.15. The time step is usually is limited by the criterion on diffusion for us, so it's on the edge.
But as I remember it, when the surface problem occured, ekm at that point became something like 10000, so I don't think this is fixable with a slightly smaller time step. Besides, dt is adjusted only at the start of a full Runge-Kutta step, and if this happens after say the first sub-step, it will already cause damage before dt is adjusted next.

**Comment by pgriewank** [comment]
I just ran into this issue when trying to run the arm_brown testcase where the initial tke in prof.inp.001 is 0 everywhere. As a quick fix I set the tke to 0.1 which seems to work.

It would be nice if the testcase could be updated in the master branch, or at least a comment put into the readme.arm_brown file to help newbies like me.

**Comment by fjansson** [comment]
@pgriewank I think the cleanest solution is to wait for Dales 4.2, which contains a fix for stable conditions (issue #30, commit 4098295e5). Then run with `sgs_surface_fix = .false.` .

```
&NAMSUBGRID
sgs_surface_fix = .false.
/
```

**Comment by pgriewank** [comment]
Thanks for the response @fjansson.

**Comment by fjansson** [comment]
Fixed in version 4.2:
Commit 4098295e5539b3 changes the phi function in the Monin-Obukhov similarity theory to behave better in stable conditions.
Commit db0c242d141b2f6 turns off the sgs_surface_fix by default. With the new phi definitions, the fix is no longer needed.
